mkosi/machine.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+from __future__ import annotations
+
+import contextlib
+import signal
+import sys
+from typing import Any, Optional, Sequence
+
+import pexpect  # type: ignore
+
+import mkosi
+
+from .backend import die
+
+
+class Machine():
+    def __init__(self, args: Optional[Sequence[str]] = None) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self.serial: pexpect.spawnu
+        self.stack = contextlib.ExitStack()
+        self.args = args
+        self.exit_code: int
+        self.verb: str
```

Comment:
We should always assign default values for all fields in a class. Otherwise, trying to access these fields before they are set will throw a non-useful exception. For serial and verb, with the current design, we'll have to use `None`.

Reply:
that's really neat, thanks!